File: ios/Classes/FlutterTXCloudPullView.m
#import "FlutterTXCloudPullView.h"

@implementation FlutterTXCloudPullViewFactory {
  NSObject<FlutterBinaryMessenger>* _messenger;
}

- (instancetype)initWithMessenger:(NSObject<FlutterBinaryMessenger>*)messenger {
  self = [super init];
  if (self) {
    _messenger = messenger;
  }
  return self;
}

- (NSObject<FlutterMessageCodec>*)createArgsCodec {
  return [FlutterStandardMessageCodec sharedInstance];
}

- (NSObject<FlutterPlatformView>*)createWithFrame:(CGRect)frame
                                   viewIdentifier:(int64_t)viewId
                                        arguments:(id _Nullable)args {
  FlutterTXCloudPullView* pullView = [[FlutterTXCloudPullView alloc] initWithFrame:frame
                                                                         viewIdentifier:viewId
                                                                              arguments:args
                                                                        binaryMessenger:_messenger];
  return pullView;
}

@end

@implementation FlutterTXCloudPullView {
  UIView* _videoView;
  TXLivePlayer* _player;

  int64_t _viewId;
  FlutterMethodChannel* _channel;
  FlutterBasicMessageChannel* _snapshotChannel;
}

- (instancetype)initWithFrame:(CGRect)frame
               viewIdentifier:(int64_t)viewId
                    arguments:(id _Nullable)args
              binaryMessenger:(NSObject<FlutterBinaryMessenger>*)messenger {
  if (self = [super init]) {
    _viewId = viewId;

    NSString* channelName = [NSString stringWithFormat:@"com.pactera.bg3cs.tencent_live_flutter/TXCloudPullView_%lld", viewId];
    _channel = [FlutterMethodChannel methodChannelWithName:channelName binaryMessenger:messenger];
    __weak __typeof__(self) weakSelf = self;
    [_channel setMethodCallHandler:^(FlutterMethodCall* call, FlutterResult result) {
      [weakSelf onMethodCall:call result:result];
    }];

    [self initSnapshotChannel:messenger widthId:viewId];

    // 创建播放器
    _player = [[TXLivePlayer alloc] init];

    // 视频画面显示
    _videoView = [[UIView alloc] initWithFrame:frame];
  }
  return self;
}

- (UIView*)view {
  return _videoView;
}

- (void)initSnapshotChannel: (NSObject<FlutterBinaryMessenger>*)messenger widthId: (int64_t)viewId{
    NSString* channelName = [NSString stringWithFormat:@"com.pactera.bg3cs.tencent_live_flutter/TXCloudPullView_Snapshot_%lld", viewId];
    _snapshotChannel = [FlutterBasicMessageChannel messageChannelWithName:channelName binaryMessenger:messenger codec:[FlutterBinaryCodec sharedInstance]];
  // 接收到flutter端snapshot调用后，将截图数据传回调给flutter层
    [_snapshotChannel setMessageHandler:^(id  _Nullable message, FlutterReply  _Nonnull callback) {
        callback(nil);
        [_player snapshot:^(TXImage *image){
            [_snapshotChannel sendMessage:UIImagePNGRepresentation(image)];
        }];
    }];
}

- (void)onMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if([[call method] isEqualToString:@"setPlayConfig"]){
    [self setPlayConfig:[call arguments] result:result];
  } else if ([[call method] isEqualToString:@"startPlay"]) {
    [self startPlay:[call arguments] result:result];
  } else if([[call method] isEqualToString:@"stopPlay"]){
    [self stopPlay:[call arguments] result:result];
  } else if([[call method] isEqualToString:@"pause"]){
    [self pause:[call arguments] result:result];
  }  else if([[call method] isEqualToString:@"resume"]){
    [self resume:[call arguments] result:result];
  }  else if([[call method] isEqualToString:@"isPlaying"]){
    [self isPlaying:[call arguments] result:result];
  } else if([[call method] isEqualToString:@"setRenderMode"]){
    [self setRenderMode:[call arguments] result:result];
  } else if([[call method] isEqualToString:@"setRenderRotation"]){
    [self setRenderRotation:[call arguments] result:result];
  } else if([[call method] isEqualToString:@"enableHardwareDecode"]){
    [self enableHardwareDecode:[call arguments] result:result];
  } else if([[call method] isEqualToString:@"setMute"]){
    [self setMute:[call arguments] result:result];
  } else if([[call method] isEqualToString:@"setVolume"]){
    [self setVolume:[call arguments] result:result];
  } else if([[call method] isEqualToString:@"setAudioRoute"]){
    [self setAudioRoute:[call arguments] result:result];
  } else if([[call method] isEqualToString:@"switchStream"]){
    [self switchStream:[call arguments] result:result];
  } else {
    result(FlutterMethodNotImplemented);
  }
}

#pragma mark - 播放基础接口

- (void) setPlayConfig:(NSDictionary<NSString*, id>*)arguments result:(FlutterResult)result {
   // NSLog(@"IOS setPlayConfig: %@", arguments);
    TXLivePlayConfig* config = _player.config;
    config.cacheTime = [arguments[@"cacheTime"] floatValue];
    config.bAutoAdjustCacheTime = [arguments[@"bAutoAdjustCacheTime"] boolValue];
    config.maxAutoAdjustCacheTime = [arguments[@"maxAutoAdjustCacheTime"] floatValue];
    config.minAutoAdjustCacheTime = [arguments[@"minAutoAdjustCacheTime"] floatValue];
    config.videoBlockThreshold = [arguments[@"videoBlockThreshold"] intValue];
    config.connectRetryCount = [arguments[@"connectRetryCount"] intValue];
    config.connectRetryInterval = [arguments[@"connectRetryInterval"] intValue];
    config.enableAEC = [arguments[@"enableAEC"] boolValue];
    config.enableMessage = [arguments[@"enableMessage"] boolValue];
    config.enableMetaData = [arguments[@"enableMetaData"] boolValue];
    config.flvSessionKey = arguments[@"flvSessionKey"];
 //   [_player setConfig:config];
    result(@(0));
}

- (void)startPlay:(NSDictionary<NSString*, id>*)arguments result:(FlutterResult)result {
    [_player setDelegate:self];
    [_player setupVideoWidget:CGRectZero containView:_videoView insertIndex:0];

    NSString* playUrl = arguments[@"playUrl"];
    TX_Enum_PlayType  _playType;               // 播放类型
    if ([playUrl hasPrefix:@"rtmp:"]) {
        _playType = PLAY_TYPE_LIVE_RTMP;
    } else if (([playUrl hasPrefix:@"https:"] || [playUrl hasPrefix:@"http:"]) && ([playUrl rangeOfString:@".flv"].length > 0)) {
        _playType = PLAY_TYPE_LIVE_FLV;
    } else if (([playUrl hasPrefix:@"https:"] || [playUrl hasPrefix:@"http:"]) && [playUrl rangeOfString:@".m3u8"].length > 0) {
        _playType = PLAY_TYPE_VOD_HLS;
    } else{
        FlutterError *error = [FlutterError errorWithCode:@"invalid url" message:@"播放地址不合法，直播目前仅支持rtmp,flv播放方式!" details:nil];
        result(error);
        return;
    }
    int ret = [_player startPlay:playUrl type:_playType];
    result(@(ret));
}

- (void)stopPlay:(NSDictionary<NSString*, id>*)arguments result:(FlutterResult)result {
    [self _stopPlay];
    result(@(0));
}

- (void)_stopPlay{
    if (_player) {
        [_player setDelegate:nil];
        [_player removeVideoWidget];
        [_player stopPlay];
    }
}

- (void)pause:(NSDictionary<NSString*, id>*)arguments result:(FlutterResult)result {
    [_player pause];
    result(@(0));
}

- (void)resume:(NSDictionary<NSString*, id>*)arguments result:(FlutterResult)result {
    [_player resume];
    result(@(0));
}

- (void)isPlaying:(NSDictionary<NSString*, id>*)arguments result:(FlutterResult)result {
    result(@([_player isPlaying]));
}

#pragma mark - 播放配置接口

- (void)setRenderMode:(NSDictionary<NSString*, id>*)arguments result:(FlutterResult)result {
    if(_player){
        int mode = [arguments[@"renderMode"] intValue];
        if (mode == 1) {
            [_player setRenderMode:RENDER_MODE_FILL_EDGE];
        } else {
            [_player setRenderMode:RENDER_MODE_FILL_SCREEN];
        }
    }
    result(@(0));
}

- (void)setRenderRotation:(NSDictionary<NSString*, id>*)arguments result:(FlutterResult)result {
    if(_player){
        int renderRotation = [arguments[@"renderRotation"] intValue];
        switch(renderRotation){
            case 0:
                [_player setRenderRotation:HOME_ORIENTATION_DOWN];
                break;
            case 90:
                [_player setRenderRotation:HOME_ORIENTATION_LEFT];
                break;
            case 180:
                [_player setRenderRotation:HOME_ORIENTATION_UP];
                break;
            case 270:
                [_player setRenderRotation:HOME_ORIENTATION_RIGHT];
                break;
            default:
                [_player setRenderRotation:HOME_ORIENTATION_DOWN];
        }
    }
    result(@(0));
}

- (void)enableHardwareDecode:(NSDictionary<NSString*, id>*)arguments result:(FlutterResult)result {
    _player.enableHWAcceleration = [arguments[@"enable"] boolValue];
    result(@(YES));
}

- (void)setMute:(NSDictionary<NSString*, id>*)arguments result:(FlutterResult)result {
    [_player setMute:[arguments[@"mute"] boolValue]];
    result(@(0));
}

- (void)setVolume:(NSDictionary<NSString*, id>*)arguments result:(FlutterResult)result {
    [_player setVolume:[arguments[@"volume"] intValue]];
    result(@(0));
}

- (void)setAudioRoute:(NSDictionary<NSString*, id>*)arguments result:(FlutterResult)result {
    int type = [arguments[@"audioRoute"] intValue];
    if(type == 0){
        [TXLivePlayer setAudioRoute:AUDIO_ROUTE_SPEAKER];
    }else if(type == 1){
        [TXLivePlayer setAudioRoute:AUDIO_ROUTE_RECEIVER];
    }
    result(@(0));
}

- (void)switchStream:(NSDictionary<NSString*, id>*)arguments result:(FlutterResult)result {
    NSString *playUrl = arguments[@"playUrl"];
    int ret = [_player switchStream:playUrl];
    result(@(ret));
}

- (void)dealloc {
    [self _stopPlay];
}

#pragma mark - TXLivePlayListener

- (void)onPlayEvent:(int)EvtID withParam:(NSDictionary *)param {
    NSMutableDictionary *dict=[NSMutableDictionary  dictionaryWithDictionary:param];
    [dict setObject:@(EvtID) forKey:@"event"];

    [_channel invokeMethod:@"onPlayEvent" arguments:dict];
}

- (void)onNetStatus:(NSDictionary *)param {

}

@end

File: android/src/main/java/com/pactera/bg3cs/tencent_live_flutter/FlutterTXCloudPullView.java
package com.pactera.bg3cs.tencent_live_flutter;

import android.app.Activity;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.platform.PlatformView;

import com.tencent.rtmp.ITXLivePlayListener;
import com.tencent.rtmp.TXLiveConstants;
import com.tencent.rtmp.TXLivePlayConfig;
import com.tencent.rtmp.TXLivePlayer;
import com.tencent.rtmp.ui.TXCloudVideoView;
import com.tencent.rtmp.TXLivePlayer.ITXSnapshotListener;

import java.util.HashMap;
import java.util.Map;

public class FlutterTXCloudPullView implements PlatformView, MethodChannel.MethodCallHandler, ITXLivePlayListener{

    private static final String TAG = "FlutterTXCloudPullView";

    private final Activity activity;
    private final MethodChannel methodChannel;
    private SnapshotChannel snapShotChannel;

    private TXCloudVideoView mPlayerView;

    // 拉流播放
    private TXLivePlayer mLivePlayer = null;
    private TXLivePlayConfig mPlayConfig;


    public FlutterTXCloudPullView(BinaryMessenger messenger, Activity activity, int id) {
        this.activity = activity;
        methodChannel = new MethodChannel(messenger, "com.pactera.bg3cs.tencent_live_flutter/TXCloudPullView_" + id);
        methodChannel.setMethodCallHandler(this);

        initSnapshotChannel(messenger, id);

        mPlayerView = new TXCloudVideoView(activity.getApplicationContext());
        initPlay();
    }

    private void initSnapshotChannel(BinaryMessenger messenger, int id){

        String channelName = "com.pactera.bg3cs.tencent_live_flutter/TXCloudPullView_Snapshot_" + id;

        snapShotChannel = new SnapshotChannel(messenger, channelName, new SnapshotChannel.Request() {
            @Override
            public void snapshotRequest() {
                mLivePlayer.snapshot(new ITXSnapshotListener(){
                    @Override
                    public void onSnapshot(Bitmap bmp){
                        snapShotChannel.sendSnapShot(bmp);
                    }
                });
            }
        });
    }

    // 必须在播放之前调用
    private void initPlay() {
        if (mLivePlayer == null) {
            mLivePlayer = new TXLivePlayer(activity.getApplicationContext());
        }
        if(mPlayConfig == null){
            mPlayConfig = new TXLivePlayConfig();
        }
    }

    // 设置配置需要放在开始播放之前，否则有部分配置无法生效
    private void setPlayConfig(Map<String, Object> request, MethodChannel.Result result){
        Log.e(TAG, request.toString());
        mPlayConfig.setCacheTime(((Double)request.get("cacheTime")).floatValue());
        mPlayConfig.setAutoAdjustCacheTime((boolean)request.get("bAutoAdjustCacheTime"));
        mPlayConfig.setMaxAutoAdjustCacheTime(((Double)request.get("maxAutoAdjustCacheTime")).floatValue());
        mPlayConfig.setMinAutoAdjustCacheTime(((Double)request.get("minAutoAdjustCacheTime")).floatValue());
        mPlayConfig.setVideoBlockThreshold((int)request.get("videoBlockThreshold"));
        mPlayConfig.setConnectRetryCount((int)request.get("connectRetryCount"));
        mPlayConfig.setConnectRetryInterval((int)request.get("connectRetryInterval"));
        mPlayConfig.enableAEC((boolean)request.get("enableAEC"));
        mPlayConfig.setEnableMessage((boolean)request.get("enableMessage"));
        mPlayConfig.setEnableMetaData((boolean)request.get("enableMetaData"));
        mPlayConfig.setFlvSessionKey((String)request.get("flvSessionKey"));
//        mLivePlayer.setConfig(mPlayConfig);
        result.success(0);
    }

    @Override
    public View getView() {
        return mPlayerView;
    }

    @Override
    public void dispose() {
        methodChannel.setMethodCallHandler(null);
        snapShotChannel.dispose();
        _stopPlay();
        if (mPlayerView != null){
            mPlayerView.onDestroy();
            mPlayerView = null;
        }
    }

    @Override
    public void onMethodCall(MethodCall methodCall, MethodChannel.Result result) {
        Map<String, Object> request = (Map<String, Object>) methodCall.arguments;
        switch (methodCall.method) {
            case "setPlayConfig":
                setPlayConfig(request, result);
                break;
            case "startPlay":
                startPlay(request, result);
                break;
            case "stopPlay":
                stopPlay(request, result);
                break;
            case "pause":
                pause(request, result);
                break;
            case "resume":
                resume(request,result);
                break;
            case "isPlaying":
                isPlaying(request, result);
                break;
            case "setRenderMode":
                setRenderMode(request, result);
                break;
            case "setRenderRotation":
                setRenderRotation(request, result);
                break;
            case "enableHardwareDecode":
                enableHardwareDecode(request, result);
                break;
            case "setMute":
                setMute(request, result);
                break;
            case "setVolume":
                setVolume(request, result);
                break;
            case "setAudioRoute":
                setAudioRoute(request, result);
                break;
            case "switchStream":
                switchStream(request, result);
                break;
            default:
                result.notImplemented();
        }
    }

    // ITXLivePlayListener
    @Override
    public void onPlayEvent(int event, Bundle param) {

        String playEventLog = "receive event: " + event + ", " + param.getString(TXLiveConstants.EVT_DESCRIPTION);
        Log.d(TAG, playEventLog);

        HashMap<String, Object> args = new HashMap<>();
        args.put("event", event);

        String desc = param.getString(TXLiveConstants.EVT_DESCRIPTION);
        if(desc != null){
            args.put(TXLiveConstants.EVT_DESCRIPTION, desc);
        }
        int defaultValue = -1111;
        int param1 = param.getInt(TXLiveConstants.EVT_PARAM1, defaultValue);
        if(param1 != defaultValue){
            args.put(TXLiveConstants.EVT_PARAM1, param1);
        }
        int param2 = param.getInt(TXLiveConstants.EVT_PARAM2, defaultValue);
        if(param2 != defaultValue){
            args.put(TXLiveConstants.EVT_PARAM2, param2);
        }

        methodChannel.invokeMethod("onPlayEvent", args);
    }

    @Override
    public void onNetStatus(Bundle status) {
//        String str = getNetStatusString(status);
//        Log.d(TAG, "Current status, CPU:" + status.getString(TXLiveConstants.NET_STATUS_CPU_USAGE) +
//                ", RES:" + status.getInt(TXLiveConstants.NET_STATUS_VIDEO_WIDTH) + "*" + status.getInt(TXLiveConstants.NET_STATUS_VIDEO_HEIGHT) +
//                ", SPD:" + status.getInt(TXLiveConstants.NET_STATUS_NET_SPEED) + "Kbps" +
//                ", FPS:" + status.getInt(TXLiveConstants.NET_STATUS_VIDEO_FPS) +
//                ", ARA:" + status.getInt(TXLiveConstants.NET_STATUS_AUDIO_BITRATE) + "Kbps" +
//                ", VRA:" + status.getInt(TXLiveConstants.NET_STATUS_VIDEO_BITRATE) + "Kbps");
//
//        mPlayVisibleLogView.setLogText(status, null, 0);
    }

    //////////////////////////////////  播放基础接口 ////////////////////////////////////

    private void startPlay(Map<String, Object> request, MethodChannel.Result result){
        String playUrl = (String) request.get("playUrl");
        int playType;
        if (playUrl != null && playUrl.startsWith("rtmp://")) {
            playType = TXLivePlayer.PLAY_TYPE_LIVE_RTMP;
        } else if (playUrl != null && (playUrl.startsWith("http://") || playUrl.startsWith("https://"))&& playUrl.contains(".flv")) {
            playType = TXLivePlayer.PLAY_TYPE_LIVE_FLV;
        } else {
            result.error("invalid url", "播放地址不合法，直播目前仅支持rtmp,flv播放方式!", null);
            return;
        }
        mLivePlayer.setPlayListener(this);
        mLivePlayer.setPlayerView(mPlayerView);
        int ret = mLivePlayer.startPlay(playUrl, playType); // result返回值：0 success;  -1 empty url; -2 invalid url; -3 invalid playType;
        result.success(ret);
    }

    private void  stopPlay(Map<String, Object> request, MethodChannel.Result result){
        _stopPlay();
        result.success(0);
    }

    private void _stopPlay(){
        if (mLivePlayer != null) {
            mLivePlayer.stopRecord();
            mLivePlayer.setPlayListener(null);
            mLivePlayer.stopPlay(true);
        }
    }

    private void pause(Map<String, Object> request, MethodChannel.Result result){
        mLivePlayer.pause();
        result.success(0);
    }

    private void resume(Map<String, Object> request, MethodChannel.Result result){
        mLivePlayer.resume();
        result.success(0);
    }

    private void isPlaying(Map<String, Object> request, MethodChannel.Result result){
        result.success(mLivePlayer.isPlaying());
    }


    //////////////////////////////////  播放配置接口 ////////////////////////////////////

    //平铺模式
    private void setRenderMode(Map<String, Object> request, MethodChannel.Result result){
        if(mLivePlayer != null){
            mLivePlayer.setRenderMode((int)request.get("renderMode"));
        }
        result.success(0);
    }

    //横屏|竖屏
    private void setRenderRotation(Map<String, Object> request, MethodChannel.Result result){
        if(mLivePlayer != null){
            mLivePlayer.setRenderRotation((int)request.get("renderRotation"));
        }
        result.success(0);
    }

    // 开启硬件加速
    private void enableHardwareDecode(Map<String, Object> request, MethodChannel.Result result){
        boolean ret = mLivePlayer.enableHardwareDecode((boolean)request.get("enable"));
        result.success(ret);
    }

    // 设置是否静音播放
    private void setMute(Map<String, Object> request, MethodChannel.Result result){
        mLivePlayer.setMute((boolean)request.get("mute"));
        result.success(0);
    }

    // 设置音量
    private void setVolume(Map<String, Object> request, MethodChannel.Result result){
        mLivePlayer.setVolume((int)request.get("volume"));
        result.success(0);
    }

    // 设置声音播放模式
    private void setAudioRoute(Map<String, Object> request, MethodChannel.Result result){
        mLivePlayer.setAudioRoute((int)request.get("audioRoute"));
        result.success(0);
    }

    // 设置声音播放模式
    private void switchStream(Map<String, Object> request, MethodChannel.Result result){
        mLivePlayer.switchStream((String)request.get("playUrl"));
        result.success(0);
    }


}
